src/ppx/geometry.cpp

```diff
@@ -13,12 +13,392 @@
+                        PPX_ASSERT_MSG(n > 0, "position should always available");
+                    }
+                    break;
+                case grfx::VERTEX_SEMANTIC_NORMAL    : AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.normal); break;
+                case grfx::VERTEX_SEMANTIC_COLOR     : AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.color); break;
+                case grfx::VERTEX_SEMANTIC_TANGENT   : AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.tangent); break;
+                case grfx::VERTEX_SEMANTIC_BITANGENT : AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.bitangent); break;
+                case grfx::VERTEX_SEMANTIC_TEXCOORD  : AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.texCoord); break;
+            }
+            // clang-format on
+        }
+        uint32_t endSize = GetVertexBufferSize(pGeom, kBufferIndex);
+
+        uint32_t       bytesWritten            = (endSize - startSize);
+        const uint32_t vertexBufferElementSize = GetVertexBufferElementSize(pGeom, kBufferIndex);
+        PPX_ASSERT_MSG(bytesWritten == vertexBufferElementSize, "size of vertex data written does not match buffer's element size");
+
+        return GetVertexBufferElementCount(pGeom, kBufferIndex);
+    }
+
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const WireMeshVertexData& vtx) override
+    {
+        uint32_t       startSize = GetVertexBufferSize(pGeom, kBufferIndex);
+        const uint32_t attrCount = GetVertexBindingAttributeCount(pGeom, kBufferIndex);
+        for (uint32_t attrIndex = 0; attrIndex < attrCount; ++attrIndex) {
+            const grfx::VertexSemantic semantic = GetVertexBindingAttributeSematic(pGeom, kBufferIndex, attrIndex);
+
+            // clang-format off
+            switch (semantic) {
+                default: break;
+                case grfx::VERTEX_SEMANTIC_POSITION: 
+                    {
+                        const uint32_t n = AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.position); 
+                        PPX_ASSERT_MSG(n > 0, "position should always available");
+                    }
+                    break;
+                case grfx::VERTEX_SEMANTIC_COLOR   : AppendDataToVertexBuffer(pGeom, kBufferIndex, vtx.color); break;
+            }
+            // clang-format on
+        }
+        uint32_t endSize = GetVertexBufferSize(pGeom, kBufferIndex);
+
+        uint32_t       bytesWritten            = (endSize - startSize);
+        const uint32_t vertexBufferElementSize = GetVertexBufferElementSize(pGeom, kBufferIndex);
+        PPX_ASSERT_MSG(bytesWritten == vertexBufferElementSize, "size of vertex data written does not match buffer's element size");
+
+        return GetVertexBufferElementCount(pGeom, kBufferIndex);
+    }
+
+    virtual uint32_t GetVertexCount(const Geometry* pGeom) override
+    {
+        return GetVertexBufferElementCount(pGeom, kBufferIndex);
+    }
+
+private:
+    const uint32_t kBufferIndex = 0;
```

Comment:
Does this refer to a specific buffer? Since the word buffer is used quite a lot in these classes can you add a comment on what this is used for?

Reply:
Oh I see, there's only one buffer in this class.